File: renderers/web/tests/access.test.js
// What a box says about itself to a reader who cannot see it.
//
// A control the platform draws says this for itself. One the engine draws is a box with a colour in it,
// and that is what a screen reader hears, so a drawn checkbox that names no role is a checkbox nobody
// using one can find, let alone tick — which is a regression against the control it replaced.

import { install } from "../dom.js";
import { WebRenderer } from "../renderer.js";

const surface = install();
const renderer = new WebRenderer(surface, () => {});

function assert(condition, message) {
    if (!condition) {
        console.error(`failed: ${message}`);
        process.exit(1);
    }
}

function box(id, props) {
    renderer.apply([
        { op: "create", id, type: "pressable", props },
        { op: "insert", id, parent: 0, index: 1 },
    ]);

    return renderer.nodes.get(id).element;
}

{
    const ticked = box(1, {
        accessibilityLabel: "Keep it",
        accessibilityRole: "checkbox",
        accessibilityState: { checked: true, disabled: false },
    });

    assert(ticked.getAttribute("role") === "checkbox", "the role reaches the page");
    assert(ticked.getAttribute("aria-label") === "Keep it", "and so does the name");
    assert(ticked.getAttribute("aria-checked") === "true", "and what it is doing");
    assert(ticked.getAttribute("aria-disabled") === "false", "including what it is not doing");

    renderer.apply([{ op: "update", id: 1, props: { accessibilityState: { checked: false } } }]);

    assert(ticked.getAttribute("aria-checked") === "false", "a state that changed is said again");
    assert(ticked.hasAttribute("aria-disabled") === false, "and one no longer given is taken off");
}

{
    const slider = box(2, {
        accessibilityRole: "slider",
        accessibilityValue: { now: 3, least: 0, most: 10 },
    });

    assert(slider.getAttribute("role") === "slider", "a slider says it is one");
    assert(slider.getAttribute("aria-valuenow") === "3", "and what it is at");
    assert(slider.getAttribute("aria-valuemin") === "0", "and where it starts");
    assert(slider.getAttribute("aria-valuemax") === "10", "and where it ends");

    // A box that is nothing in particular carries no role rather than one nobody reads.
    renderer.apply([{ op: "update", id: 2, props: { accessibilityRole: "none" } }]);

    assert(slider.hasAttribute("role") === false, "a box that is nothing carries no role");
}

{
    // A role the browser spells differently is written the way the browser spells it.
    const searching = box(3, { accessibilityRole: "search" });

    assert(searching.getAttribute("role") === "searchbox",
        `a search box is a searchbox on the page, it was ${searching.getAttribute("role")}`);
}

console.log("web.access ok");

File: renderers/android/src/test/kotlin/dev/varn/gui/AccessTest.kt
package dev.varn.gui

import android.app.Activity
import android.view.View
import android.view.accessibility.AccessibilityNodeInfo
import android.widget.FrameLayout
import org.json.JSONArray
import org.json.JSONObject
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.Robolectric
import org.robolectric.RobolectricTestRunner

/**
 * What a box says about itself to a reader who cannot see it.
 *
 * A control the platform draws says this for itself. One the engine draws is a box with a colour in it,
 * and that is what TalkBack reads, so a drawn checkbox that names no role is a checkbox nobody using one
 * can find, let alone tick.
 */
@RunWith(RobolectricTestRunner::class)
class AccessTest {
    private lateinit var surface: FrameLayout
    private lateinit var renderer: VarnRenderer

    @Before
    fun setUp() {
        val activity = Robolectric.buildActivity(Activity::class.java).setup().get()

        surface = FrameLayout(activity)
        activity.setContentView(surface)

        renderer = VarnRenderer(activity, surface) { _, _, _ -> }
    }

    private fun box(props: Map<String, Any?>): View {
        renderer.apply(
            JSONArray(
                listOf(
                    JSONObject(mapOf("op" to "create", "id" to 1, "type" to "pressable", "props" to JSONObject(props))),
                    JSONObject(mapOf("op" to "insert", "id" to 1, "parent" to 0, "index" to 1)),
                ),
            ),
        )

        return surface.getChildAt(0)
    }

    private fun spoken(view: View): AccessibilityNodeInfo {
        val info = AccessibilityNodeInfo.obtain()

        view.onInitializeAccessibilityNodeInfo(info)
        return info
    }

    @Test
    fun aDrawnControlSaysWhatItIsAndWhatItIsDoing() {
        val view = box(
            mapOf(
                "accessibilityLabel" to "Keep it",
                "accessibilityRole" to "checkbox",
                "accessibilityState" to JSONObject(mapOf("checked" to true)),
            ),
        )

        val info = spoken(view)

        assertEquals("Keep it", view.contentDescription)
        assertEquals("android.widget.CheckBox", info.className)
        assertTrue("a checkbox says it can be ticked", info.isCheckable)
        assertTrue("and that it is", info.isChecked)
    }

    @Test
    fun aControlThatCannotBeUsedSaysSo() {
        val view = box(
            mapOf(
                "accessibilityRole" to "button",
                "accessibilityState" to JSONObject(mapOf("disabled" to true)),
            ),
        )

        assertFalse("a control the tree disabled is not one a reader can reach", view.isEnabled)
    }

    @Test
    fun aControlThatHoldsANumberSaysWhatItIsAt() {
        val view = box(
            mapOf(
                "accessibilityRole" to "slider",
                "accessibilityValue" to JSONObject(mapOf("now" to 3, "least" to 0, "most" to 10)),
            ),
        )

        val range = spoken(view).rangeInfo

        assertEquals(3.0f, range?.current ?: -1f, 0.001f)
        assertEquals(0.0f, range?.min ?: -1f, 0.001f)
        assertEquals(10.0f, range?.max ?: -1f, 0.001f)
    }

    @Test
    fun aBoxThatIsNothingInParticularSaysNothing() {
        val view = box(mapOf("accessibilityRole" to "none"))

        assertEquals(
            "a box that is nothing is not held open for a reader",
            View.IMPORTANT_FOR_ACCESSIBILITY_AUTO,
            view.importantForAccessibility,
        )
    }
}
